subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/AbstractLazyModuleComponentResolveMetadata.java

```diff
@@ -188,6 +189,23 @@ private void populateHierarchy(Configuration metadata, ImmutableSet.Builder<Stri
         }
     }
 
+    private ImmutableSet<String> constructVariantHierarchy(ConfigurationMetadata base, AdditionalVariant additionalVariant) {
```

Comment:
I believe hierarchy and carrying over excludes should be enough.

I tried this:

```
    private ImmutableList<ExcludeMetadata> constructVariantExcludes(ConfigurationMetadata base, AdditionalVariant additionalVariant) {
        if(base == null || base.getExcludes().isEmpty()) {
            return ImmutableList.of();
        }


        ImmutableList.Builder<ExcludeMetadata> excludeMetadata = ImmutableList.builder();

        base.getExcludes().forEach((baseExclude) -> {
            if(baseExclude instanceof DefaultExclude) {
                Set<String> configurationsToExclude = new HashSet<>(((DefaultExclude) baseExclude).getConfigurations());
                configurationsToExclude.add(additionalVariant.getName());
                excludeMetadata.add(new DefaultExclude(baseExclude.getModuleId(), baseExclude.getArtifact(), configurationsToExclude.stream().toArray(String[] ::new), baseExclude.getMatcher()));
            }
        });

        return excludeMetadata.build();
    }
``` 

However, I don't think it is really helpful as `runtimeElements` and `apiElements` won't be configurations at least not in ivy but it hierarchy is reflected well: `compile` `apiElements`, excludes for `compile` should be applied to `apiElements`. 

Let me know your thoughts on this